File: csharp/yacht/Yacht.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum YachtCategory
{
    Ones = 1,
    Twos = 2,
    Threes = 3,
    Fours = 4,
    Fives = 5,
    Sixes = 6,
    FullHouse = 7,
    FourOfAKind = 8,
    LittleStraight = 9,
    BigStraight = 10,
    Choice = 11,
    Yacht = 12,
}

public static class YachtGame
{
    public static int Score(int[] dice, YachtCategory category)
    {
        switch (category)
        {
            case YachtCategory.Yacht:
                return dice.IsYacht();
            case YachtCategory.Ones:
            case YachtCategory.Twos:
            case YachtCategory.Threes:
            case YachtCategory.Fours:
            case YachtCategory.Fives:
            case YachtCategory.Sixes:
                return dice.IsAnyCombinationOf(category);
            case YachtCategory.FullHouse:
                return dice.IsFullHouse();
            case YachtCategory.FourOfAKind:
                return dice.IsFourOfAKind();
            case YachtCategory.LittleStraight:
                return dice.IsLittleStraight();
            case YachtCategory.BigStraight:
                return dice.IsBigStraight();
            case YachtCategory.Choice:
                return dice.Sum();
            default:
                return 0;
        }
    }

    private static int IsYacht(this int[] dice)
        => dice.Distinct().Count() == 1
            ? 50
            : 0;

    private static int IsAnyCombinationOf(this int[] dice, YachtCategory yachtCategory)
        => dice.Where(value => value == (int)yachtCategory)
               .Sum();

    private static int IsFullHouse(this int[] dice)
        => dice.GroupBy(value => value)
               .All(diceGroup => diceGroup.Count() == 2 || diceGroup.Count() == 3)
            ? dice.Sum()
            : 0;

    private static int IsFourOfAKind(this int[] dice)
        => dice.GroupBy(value => value)
               .FirstOrDefault(value => value.Count() >= 4)
               ?.First() * 4 ?? 0;

    private static int IsLittleStraight(this int[] dice)
        => dice.OrderBy(value => value).SequenceEqual(Enumerable.Range(1, 5))
            ? 30
            : 0;

    private static int IsBigStraight(this int[] dice)
        => dice.OrderBy(value => value).SequenceEqual(Enumerable.Range(2, 5))
            ? 30
            : 0;
}

File: fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let value die =
    match die with
    | One -> 1
    | Two -> 2
    | Three -> 3
    | Four -> 4
    | Five -> 5
    | Six -> 6

let private straightPoints = 30

let private yachtPoints = 50

let private countsOf =
    List.countBy id >> List.map snd >> List.sort

let private countBy die =
    List.filter ((=) die) >> List.length

let private total =
    List.fold (fun state d -> state + value d) 0

let score (category: Category) (dice: Die list) =
    match (category, countsOf dice) with
    | (Ones, _) ->
        (dice |> countBy Die.One) * value Die.One
    | (Twos, _) ->
        (dice |> countBy Die.Two) * value Die.Two
    | (Threes, _) ->
        (dice |> countBy Die.Three) * value Die.Three
    | (Fours, _) ->
        (dice |> countBy Die.Four) * value Die.Four
    | (Fives, _) ->
        (dice |> countBy Die.Five) * value Die.Five
    | (Sixes, _) ->
        (dice |> countBy Die.Six) * value Die.Six
    | (FullHouse, [2;3]) ->
        total dice
    | (FourOfAKind, [1;4]) | (FourOfAKind, [5]) ->
        dice
        |> List.countBy id
        |> List.find (fun (_, count) -> count >= 4)
        |> fun (die, _) -> value die * 4
    | (LittleStraight, [1;1;1;1;1]) when dice |> countBy Die.Six = 0 ->
        straightPoints
    | (BigStraight, [1;1;1;1;1]) when dice |> countBy Die.One = 0 ->
        straightPoints
    | (Choice, _) ->
        total dice
    | (Yacht, [5]) ->
        yachtPoints
    | _ -> 0